### DISCUSSION TITLE
Optimization for a large path of convertions

### DISCUSSION THREAD

**Discussion by maheca** [ImageMagick/ImageMagick#2453] [Help]
Hi all, I´m working on a project based in a Banana Pi, with Armbian in server mode (no GUI). All project is writed in c.

I´m getting a yuyv422 image from a web camera, and my final goal is to obtain data from QR codes that are near from the camera.

Everything works fine, but now, I´m trying to do it faster (Total time between code appearing on camera and I get decoded data are about 2 seconds.

My complete chain of conversion is:

Step 1: Get camera streaming (yuyv422 640x480 @16fps)

Step 2: Decode frame using FFMPEG C-API

Step 3: Create a temporary .bmp file (/tmp/pthoto.bmp) based on converted frame (RGB - 24 bits per pixel)

Step 4: Use system command "convert" inside my c program

`system("convert` /tmp/photo.bmp -set colorspace Gray -separate -average /tmp/temporaryPhoto.png");`

Step 5: Use libpng to get data from /tmp/temporaryPhoto.png file

Step 6: Use zlib to get decoded data

So, please I would to know if somebody could suggest some way to avoid (or optimize) some of this steps. (For example, I guess if there is a way to avoid writing files, working directly from data buffers. Would be fantastic if zlib could work with bmp files, but AFAIK this is not an option (please correct me if I´m wrong).

Thanks in advances to all of you,

Regards.

**Comment by fmw42** [comment]
Why are you setting colorspace to gray and then averaging the channels? When you separate channels, the results will be gray. So setting colorspace to gray does nothing. If you want the equal average of the channels of the input, you can do

`convert image.bmp -grayscale average result.png`

See https://imagemagick.org/script/command-line-options.php#grayscale and https://imagemagick.org/script/command-line-options.php#intensity

You can avoid writing to file for your output by specifying the output format in your stdout as

`convert image.bmp -grayscale average PNG:- | some other commands not part of ImageMagick.`

You may be able to read your YUV422 format directly with ImageMagick. See https://imagemagick.org/script/formats.php. But I am not an expert on that.

**Reply by maheca** [reply]
Hi fmw42, Ok, I got it. Sincerelly I´ve obtained an example of use, and maybe I.ve used it wrongly, I´m not an expert in imagemagick (as you can see). Thank you very much for your observation.

I´ll try your suggestions, and I´ll be back with results.

One more time, thanks!

**Comment by snibgo** [comment]
You have:
```
-set colorspace Gray -separate -average
```
... but "-set colorspace Gray" will convert the image to a single channel, representing gray. Then "-separate" and "-average" will do nothing.

You could link IM with your C program, and call IM functions, instead of a system call. But if your BMP file is not compressed, with RGB channels, the C code to read that and average the channels at each pixel is trivial, so there would be no need to use IM.

Writing a PNG file is non-trivial (you would call libpng to do that) but is that step needed? I don't know what step 6 "Use zlib to get decoded data" is for. Perhaps that is in a program that reads the QR data.

**Reply by maheca** [reply]
Hi snibgo, ok, I must have misschanged some example obtained on the web, thanks for your observation.

As you said, step 6 is needed to obtain decoded data from an image. Unfortunatelly, if I´m not wrong, zlib just support png files as source images, so I need to take yuyv422 data and convert it to write on my frame buffer (10,1" HDMI display, RGBA), and then convert image data to a gray png, then pass it to zlib.

I´ll study your suggestions.

Thank you very much.
